### DISCUSSION TITLE
Cycle Start Time Not working

### DISCUSSION THREAD

**Discussion by Whynolove** [mwood77/winderoo#10] [Q&A] [answered]
I've just put everything together and the winder works with the control buttons. However, the "Cycle Start Time" function doesn't seem to be functioning. I have it enabled, but it does not start the winder at the designated time. I've triple checked my wiring and made sure everything else was correct. I think it may be a software issue? Please help.

**Comment by mwood77** [comment]
Hi mate, apologies for not answering sooner. I didn’t see a notification for this.

That’s interesting. What time did you have it set for? I noticed a bug in the code for the cycle start timer earlier this week, and pushed a fix for it yesterday.

Besides that, it could be that it isn’t detecting your locale correctly. It determines your time offset with an IP geo lookup… could’ve gone wrong there. 

It highly likely it’s a software bug, as there’s no wiring dependency for this to work.

* First question, which ESP32 did you buy? Do you have a link?

<hr>

This is where it gets your local time and syncs itself; the ESP32 has a notoriously inaccurate RTC:
https://github.com/mwood77/winderoo/blob/3895e7565e6fdde638afeccf96166e3ba4f5d53e/src/platformio/osww-server/src/main.cpp#L122-L139

- Second question, can you make sure the `timezone` and `raw_offset` fields are correct for your locale, when you click on this link:
http://worldtimeapi.org/api/ip

<hr>

Lastly, this is where it determines whether to begin winding, when the timer is enabled:
https://github.com/mwood77/winderoo/blob/3895e7565e6fdde638afeccf96166e3ba4f5d53e/src/platformio/osww-server/src/main.cpp#L537-L546

In the above, it’s asking
1. Is the timer enabled?
2. Is the current hour == to the alarm hour desired?
3. Is the current minute == to the alarm minute desired?
4. Is the routine **not currently running**?
5. Is the winder enabled? (That’s the on/off toggle at the top of the screen - it must be set to `ON`)

If all of these evaluate correctly, the winder should begin at your desired time.

**Comment by Whynolove** [comment]
Hey,

Thanks for the reply. I actually figured it out. It was the latter: the program seems to think it's in a different time zone. It is 17 hours ahead for some reason. Right now I've just adjusted the time accordingly and it's worked for the past 3 days.

To use your new code, should I re-flash the board?

________________________________

**Comment by mwood77** [comment]
Are you behind a VPN?

<hr>

To fix the bug, pull the new code, then re-flash the code & file system to the board. The bug was in the frontend code, which is stored in the filesystem.

**Comment by Whynolove** [comment]
Nope, no VPN.

What was the bugfix?
________________________________

**Comment by mwood77** [comment]
When selecting the hour `23`, it would show as `3`.

**Comment by mwood77** [comment, marked as answer]
This was an issue with instantiating the rtc using an epoch/offset. I've moved this to using local datestamp time in #49
